ntp-proto/src/filter.rs

```diff
@@ -451,6 +444,116 @@ fn find_interval(chime_list: &[CandidateTuple]) -> Option<(NtpDuration, NtpDurat
+
+    let mut min_peer_jitter = 2.0e9;
+    let mut max_selection_jitter = -2.0e9;
+
+    loop {
+        // for each candidate
+        for (index, candidate) in candidates.iter().enumerate() {
+            let p = candidate.peer;
+
+            min_peer_jitter = p.statistics.jitter.min(min_peer_jitter);
+
+            // compute the selection jitter
+            let selection_jitter = candidates
+                .iter()
+                .map(|q| p.statistics.offset - q.peer.statistics.offset)
+                .map(|v| v.to_seconds().powi(2))
+                .sum::<f64>()
+                .sqrt();
+
+            if selection_jitter > max_selection_jitter {
+                qmax_index = index;
+                max_selection_jitter = selection_jitter;
+            }
+        }
+
+        // If the maximum selection jitter is less than the
+        // minimum peer jitter, then tossing out more survivors
+        // will not lower the minimum peer jitter, so we might
+        // as well stop.  To make sure a few survivors are left
+        // for the clustering algorithm to chew on, we also stop
+        // if the number of survivors is less than or equal to
+        // NMIN (3).
+        if max_selection_jitter < min_peer_jitter || candidates.len() <= MIN_CLUSTER_SURVIVORS {
+            return;
+        }
+
+        // delete the survivor qmax (the one with the highest jitter) and go around again
+        candidates.remove(qmax_index);
+    }
+}
+
+#[allow(dead_code)]
+fn clock_select(
+    peers: &[Peer],
+    local_clock_time: NtpTimestamp,
+    system_poll: NtpDuration,
+) -> Option<Vec<SurvivorTuple>> {
+    let valid_associations = peers
+        .iter()
+        .filter(|p| p.accept_synchronization(local_clock_time, system_poll));
+
+    let candidates = construct_candidate_list(valid_associations, local_clock_time);
+
+    let mut survivors = construct_survivors(&candidates, local_clock_time);
+
+    if survivors.len() < MIN_INTERSECTION_SURVIVORS {
+        return None;
+    }
+
+    cluster_algorithm(&mut survivors);
+
+    Some(survivors)
+}
+
+#[allow(dead_code)]
+struct ClockCombine {
+    offset: NtpDuration,
+    jitter: NtpDuration,
+}
+
+/// Combine the offsets of the clustering algorithm survivors
+/// using a weighted average with weight determined by the root
+/// distance.  Compute the selection jitter as the weighted RMS
+/// difference between the first survivor and the remaining
+/// survivors.  In some cases, the inherent clock jitter can be
+/// reduced by not using this algorithm, especially when frequent
+/// clockhopping is involved.  The reference implementation can
+/// be configured to avoid this algorithm by designating a
+/// preferred peer.
+#[allow(dead_code)]
+fn clock_combine<'a>(
+    survivors: &'a [SurvivorTuple<'a>],
+    local_clock_time: NtpTimestamp,
+) -> ClockCombine {
+    let mut y = 0.0;
+    let mut z = 0.0;
+    let mut w = 0.0;
+
+    let first_offset = survivors[0].peer.statistics.offset;
```

Comment:
Who is responsible for the sorting that needs to be done before this point?

Reply:
good point. it is implied that this function only runs after the clustering algorithm (which gives the correctly-sorted output). But the code skeleton does not actually enforce this ...

Reply:
FWIW, this might be a good use case for wrapping a newtype around it that can be used to assert it's sorted?